contracts/src/LightClient.sol

```diff
@@ -0,0 +1,157 @@
+    struct LightClientState {
+        uint64 viewNum;
+        uint64 blockHeight;
+        uint256 blockComm;
+        uint256 feeLedgerComm;
+        uint256 stakeTableBlsKeyComm;
+        uint256 stakeTableSchnorrKeyComm;
+        uint256 stakeTableAmountComm;
+        uint256 threshold;
+    }
+
+    /// @notice Event that a new finalized state has been successfully verified and updated
+    event NewState(uint64 indexed viewNum, uint64 indexed blockHeight, uint256 blockComm);
+
+    /// @notice The state is outdated and older than currently known `finalizedState`
+    error OutdatedState();
+    /// @notice Warning that the last block of the current epoch is not yet submitted before newer
+    /// states are proposed.
+    error MissingLastBlockForCurrentEpoch(uint64 expectedBlockHeight);
+    /// @notice Invalid user inputs: wrong format or non-sensible arguments
+    error InvalidArgs();
+
+    constructor(LightClientState memory genesis, address stakeTableAddr) {
+        if (genesis.viewNum != 0 || genesis.blockHeight != 0) {
+            revert InvalidArgs();
+        }
+
+        genesisState = genesis;
+        finalizedState = genesis;
+        currentEpoch = 0;
+        stakeTable = IStakeTable(stakeTableAddr);
+        // TODO: (alex) initialized stake table or at least store its contract address ref here
+    }
+
+    // === State Modifying APIs ===
+    //
+    /// @notice Update the latest finalized light client state. It is expected to be updated
+    /// periodically, especially an update for the last block for every epoch has to be submitted
+    /// before any newer state can be accepted since the stake table commitments of that block
+    /// become the snapshots used for vote verifications later on.
+    function newFinalizedState(
+        LightClientState calldata newState,
+        IPlonkVerifier.PlonkProof calldata proof
+    ) external {
+        if (
+            newState.viewNum <= finalizedState.viewNum
+                || newState.blockHeight <= finalizedState.blockHeight
+        ) {
+            revert OutdatedState();
+        }
+        uint64 epochEndingBlockHeight = (currentEpoch + 1) * BLOCKS_PER_EPOCH - 1;
+        if (
+            finalizedState.blockHeight != epochEndingBlockHeight
+                && newState.blockHeight > epochEndingBlockHeight
+        ) {
+            revert MissingLastBlockForCurrentEpoch(epochEndingBlockHeight);
+        }
+        // format validity check
+        BN254.validateScalarField(newState.blockComm);
+        BN254.validateScalarField(newState.feeLedgerComm);
+        BN254.validateScalarField(newState.stakeTableBlsKeyComm);
+        BN254.validateScalarField(newState.stakeTableSchnorrKeyComm);
+        BN254.validateScalarField(newState.stakeTableAmountComm);
+        // sanity check on the threshold
+        if (newState.threshold <= stakeTable.totalVotingStake() * 2 / 3) {
```

Comment:
I think we should remove this check:
* The stake table contract is no longer doing snapshots, so this would be the voting stake now, possibly mid-epoch
* If we are going to enforce that `threshold` is 2/3 + 1, why store threshold at all? Why not just compute it?
Instead, I think we can just use the previous threshold, which has been signed by a quorum of replicas from the previous epoch. And then update the threshold based on `newState` when the epoch changes.

Reply:
good point, done in [6ef422c](https://github.com/EspressoSystems/espresso-sequencer/pull/754/commits/6ef422c33121d8566894d772413543c26142ef9a)